server/handlers.go

```diff
@@ -1026,6 +1054,88 @@ func (s *Server) handleRefreshToken(w http.ResponseWriter, r *http.Request, clie
 	s.writeAccessToken(w, idToken, accessToken, rawNewToken, expiry)
 }
 
+func (s *Server) handleUserInfo(w http.ResponseWriter, r *http.Request) {
+	authorization := r.Header.Get("Authorization")
+	parts := strings.Fields(authorization)
+
+	if len(parts) != 2 || !strings.EqualFold(parts[0], "bearer") {
+		msg := "invalid authorization header"
+		w.Header().Set("WWW-Authenticate", fmt.Sprintf(`Bearer realm="dex", error="%s", error_description="%s"`, errInvalidRequest, msg))
+		s.tokenErrHelper(w, errInvalidRequest, msg, http.StatusBadRequest)
+		return
+	}
+
+	token := parts[1]
+
+	verified, err := s.verify(token)
+	if err != nil {
+		if err == errTokenExpired {
+			s.tokenErrHelper(w, errAccessDenied, err.Error(), http.StatusUnauthorized)
+			return
+		}
+		s.tokenErrHelper(w, errInvalidRequest, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(verified)
+}
+
+func (s *Server) verify(token string) ([]byte, error) {
+	keys, err := s.storage.GetKeys()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get keys: %v", err)
+	}
+
+	if keys.SigningKey == nil {
+		return nil, fmt.Errorf("no private keys found")
+	}
+
+	object, err := jose.ParseSigned(token)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse signed message")
+	}
+
+	// Parse the message to check expiry, as it jose doesn't distinguish expiry error from others
+	parts := strings.Split(token, ".")
+	if len(parts) != 3 {
+		return nil, fmt.Errorf("square/go-jose: compact JWS format must have three parts")
+	}
+
+	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: check other claims
```

Comment:
This could be a "big" TODO. I realised later on that the jose or any other JWT token lib should do all the checks, but I'm not sure they do all the needed. Or maybe they do, and I'm just being over paranoid here to check again. Need a JWT expert or jose expert here to answer the question.